### PR TITLE
update analyze_results.py

### PR DESCRIPTION
**Pull request by wanghonglie**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Updated analyze_result.py

## Modification

Updated analyze_result.py

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: tools/analysis_tools/analyze_results.py [hunk 1]
```diff
@@ -340,17 +364,19 @@ def main():
...
-        cfg.data.test.pipeline = get_loading_pipeline(
-            cfg.data.train.dataset.pipeline)
+    cfg.test_dataloader.dataset.test_mode = True
+
+    cfg.test_dataloader.pop('batch_size', 0)
+    if cfg.train_dataloader.dataset.type in ('MultiImageMixDataset',
+                                             'ClassBalancedDataset',
+                                             'RepeatDataset', 'ConcatDataset'):
+        cfg.test_dataloader.dataset.pipeline = get_loading_pipeline(
+            cfg.train_dataloader.dataset.pipeline)
```
Comment (chhluo) [inline tools/analysis_tools/analyze_results.py:374]: `get_loading_pipeline(cfg.train_dataloader.dataset.dataset.pipeline)`, see https://github.com/open-mmlab/mmdetection/blob/1fc43df01639c1dca84d5bc9ba2343eae888250f/configs/strong_baselines/mask_rcnn_r50_caffe_fpn_syncbn-all_rpn-2conv_lsj_100e_coco.py#L72

File: tools/analysis_tools/analyze_results.py [hunk 2]
```diff
@@ -340,17 +364,19 @@ def main():
...
     else:
-        cfg.data.test.pipeline = get_loading_pipeline(cfg.data.train.pipeline)
+        cfg.test_dataloader.dataset.pipeline = get_loading_pipeline(
+            cfg.train_dataloader.dataset.pipeline)
 
-    dataset = build_dataset(cfg.data.test)
+    dataset = DATASETS.build(cfg.test_dataloader.dataset)
     outputs = mmcv.load(args.prediction_path)
 
     result_visualizer = ResultVisualizer(args.show, args.wait_time,
```
Comment (chhluo) [inline tools/analysis_tools/analyze_results.py:382]: Please check whether [create_result_gif](https://github.com/open-mmlab/mmdetection/blob/test-3.0.0rc0/demo/create_result_gif.py) need modification.

File: tools/analysis_tools/analyze_results.py [hunk 3]
```diff
@@ -217,6 +240,7 @@ def detection_evaluate(self, dataset, results, topk=20, eval_fn=None):
 
         return good_mAPs, bad_mAPs
 
+    # TODO: Waiting for panoptic segmentation reconstruction to test.
```
Comment (chhluo) [inline tools/analysis_tools/analyze_results.py:243]: panoptic segmentation reconstruction is done

File: configs/_base_/datasets/coco_panoptic.py [hunk 4]
```diff
@@ -53,32 +53,35 @@
...
-#         data_root=data_root,
-#         ann_file='annotations/panoptic_image_info_test-dev2017.json',
-#         data_prefix=dict(img='test2017/', seg=None),
-#         test_mode=True,
-#         pipeline=test_pipeline))
-# test_evaluator = dict(
-#     type='CocoPanopticMetric',
-#     format_only=True,
-#     outfile_prefix='./work_dirs/coco_panoptic/test')
+test_dataloader = dict(
```
Comment (ZwwWayne) [inline configs/_base_/datasets/coco_panoptic.py:68]: we should not modify this config

File: mmdet/utils/misc.py [hunk 5]
```diff
@@ -5,6 +5,7 @@
 import warnings
 
 import mmcv
+import mmengine
 from mmcv.utils import print_log
```
Comment (RangiLyu) [inline mmdet/utils/misc.py:9]:
```suggestion
from mmengine.logging import print_log
```

File: mmdet/utils/misc.py [hunk 6]
```diff
@@ -52,8 +53,8 @@ def update_data_root(cfg, logger=None):
         cfg (mmcv.Config): The model config need to modify
         logger (logging.Logger | str | None): the way to print msg
     """
-    assert isinstance(cfg, mmcv.Config), \
-        f'cfg got wrong type: {type(cfg)}, expected mmcv.Config'
+    assert isinstance(cfg, mmengine.Config), \
```
Comment (chhluo) [inline mmdet/utils/misc.py:56]: from mmengine.config import Config

File: tools/analysis_tools/analyze_results.py [hunk 7]
```diff
@@ -110,33 +121,49 @@ def _save_image_gts_results(self,
...
-                data_info,
-                results[index],
-                dataset.CLASSES,
-                gt_bbox_color=dataset.PALETTE,
-                gt_text_color=(200, 200, 200),
-                gt_mask_color=dataset.PALETTE,
-                det_bbox_color=dataset.PALETTE,
-                det_text_color=(200, 200, 200),
-                det_mask_color=dataset.PALETTE,
+                gt_samples,
```
Comment (chhluo) [inline tools/analysis_tools/analyze_results.py:162]: visualizer will be update in  https://github.com/open-mmlab/mmdetection/pull/8620

File: tools/analysis_tools/analyze_results.py [hunk 8]
```diff
@@ -340,22 +382,34 @@ def main():
...
+                                             'RepeatDataset', 'ConcatDataset'):
+        cfg.test_dataloader.dataset.pipeline = get_loading_pipeline(
+            cfg.train_dataloader.dataset.dataset.pipeline)
     else:
-        cfg.data.test.pipeline = get_loading_pipeline(cfg.data.train.pipeline)
+        cfg.test_dataloader.dataset.pipeline = get_loading_pipeline(
+            cfg.train_dataloader.dataset.pipeline)
 
-    dataset = build_dataset(cfg.data.test)
+    dataset = DATASETS.build(cfg.test_dataloader.dataset)
```
Comment (chhluo) [inline tools/analysis_tools/analyze_results.py:397]: from mmengine.fileio import load, dump

File: tools/analysis_tools/analyze_results.py [hunk 9]
```diff
@@ -237,38 +290,26 @@ def panoptic_evaluate(self, dataset, results, topk=20):
...
-
-        result_files, tmp_dir = dataset.format_results(results)
-        pred_json = mmcv.load(result_files['panoptic'])['annotations']
-        pred_folder = osp.join(tmp_dir.name, 'panoptic')
-        gt_folder = dataset.seg_prefix
+        # gt_json = dataset.coco.img_ann_map
 
         pqs = {}
         prog_bar = mmcv.ProgressBar(len(results))
+
```
Comment (chhluo) [inline tools/analysis_tools/analyze_results.py:297]: ProgressBar is moved to mmengine now, please rebase on the lastest dev-3.x

File: mmdet/evaluation/metrics/coco_panoptic_metric.py [hunk 10]
```diff
@@ -482,7 +482,6 @@ def compute_metrics(self, results: list) -> Dict[str, float]:
             pred_json = load(json_filename)
             pred_json = dict(
                 (el['image_id'], el) for el in pred_json['annotations'])
-
```
Comment (chhluo) [inline mmdet/evaluation/metrics/coco_panoptic_metric.py:485]: unnecessary to delete this blank line

File: tools/analysis_tools/analyze_results.py [hunk 11]
```diff
@@ -341,22 +384,33 @@ def main():
...
-                               'RepeatDataset', 'ConcatDataset'):
-        cfg.data.test.pipeline = get_loading_pipeline(
-            cfg.data.train.dataset.pipeline)
+    cfg.test_dataloader.dataset.test_mode = True
+
+    cfg.test_dataloader.pop('batch_size', 0)
+    if cfg.train_dataloader.dataset.type in ('MultiImageMixDataset',
+                                             'ClassBalancedDataset',
+                                             'RepeatDataset', 'ConcatDataset'):
+        cfg.test_dataloader.dataset.dataset.pipeline = get_loading_pipeline(
```
Comment (chhluo) [inline tools/analysis_tools/analyze_results.py:393]: Why `cfg.test_dataloader.dataset.dataset.pipeline` ?

test dataset will not be a dataset like 'MultiImageMixDataset',                                       'ClassBalancedDataset',  RepeatDataset', 'ConcatDataset'

File: tools/analysis_tools/analyze_results.py [hunk 12]
```diff
@@ -237,39 +292,26 @@ def panoptic_evaluate(self, dataset, results, topk=20):
...
+
+            self.evaluator.process([data_sample])
+            metrics = self.evaluator.evaluate(1)
+
+            pqs[i] = metrics['coco_panoptic/PQ']
             prog_bar.update()
 
-        if tmp_dir is not None:
-            tmp_dir.cleanup()
+        # if tmp_dir is not None:
```
Comment (chhluo) [inline tools/analysis_tools/analyze_results.py:313]: remove it

File: tools/analysis_tools/analyze_results.py [hunk 13]
```diff
@@ -237,39 +292,26 @@ def panoptic_evaluate(self, dataset, results, topk=20):
...
-                dataset.categories,
-                dataset.file_client,
-                print_log=False)
-            pq_results, classwise_results = pq_stat.pq_average(
-                dataset.categories, isthing=None)
-            pqs[i] = pq_results['pq']
+            data_sample = {}
+            for k in dataset[i].keys():
+                data_sample[k] = dataset[i][k]
+            # data_batch = {}
```
Comment (chhluo) [inline tools/analysis_tools/analyze_results.py:302]: remove it

File: tools/analysis_tools/analyze_results.py [hunk 14]
```diff
@@ -84,17 +89,22 @@ def __init__(self,
                  show=False,
                  wait_time=0,
                  score_thr=0,
-                 overlay_gt_pred=False):
+                 overlay_gt_pred=False,
```
Comment (chhluo) [inline tools/analysis_tools/analyze_results.py:92]: `DetLocalVisualizer` only support `overlay_gt_pred=True` now, see https://github.com/open-mmlab/mmdetection/blob/e3c0b6946f33ffefd6eb52b93ea5fd015dda4281/mmdet/visualization/local_visualizer.py#L378

So `overlay_gt_pred` may be unnecessary now.

File: tools/analysis_tools/analyze_results.py [hunk 15]
```diff
@@ -111,33 +110,53 @@ def _save_image_gts_results(self,
...
+
+                data_samples = DetDataSample()
+                data_samples.pred_panoptic_seg = pred_panoptic_seg
+                data_samples.gt_panoptic_seg = gt_panoptic_seg
+
+            self.visualizer.add_datasample(
+                'image',
+                results[index]['img'],
+                data_samples,
+                show=True,
```
Comment (chhluo) [inline tools/analysis_tools/analyze_results.py:155]: update

File: tools/analysis_tools/analyze_results.py [hunk 16]
```diff
@@ -157,25 +176,32 @@ def evaluate_and_show(self,
...
         else:
-            raise 'The format of result is not supported yet. ' \
-                'Current dict for panoptic segmentation and list ' \
-                'for object detection are supported.'
-
-        good_dir = osp.abspath(osp.join(show_dir, 'good'))
-        bad_dir = osp.abspath(osp.join(show_dir, 'bad'))
-        self._save_image_gts_results(dataset, results, good_samples, good_dir)
-        self._save_image_gts_results(dataset, results, bad_samples, bad_dir)
+            raise 'expect \'pred_panoptic_seg\' or \'pred_panoptic_seg\' \
```
Comment (chhluo) [inline tools/analysis_tools/analyze_results.py:203]: expect \'pred_panoptic_seg\' or \'pred_instances\'

File: tools/analysis_tools/analyze_results.py [hunk 17]
```diff
@@ -327,6 +354,7 @@ def parse_args():
 
 
 def main():
+    register_all_modules()
```
Comment (chhluo) [inline tools/analysis_tools/analyze_results.py:357]: remove 
image

File: tools/analysis_tools/analyze_results.py [hunk 18]
```diff
@@ -73,28 +78,22 @@ class ResultVisualizer:
...
-            The image above is drawn with gt, and the image below is drawn
-            with the prediction result. Default: False.
     """
 
-    def __init__(self,
-                 show=False,
-                 wait_time=0,
-                 score_thr=0,
-                 overlay_gt_pred=False):
+    def __init__(self, show=False, wait_time=0, score_thr=0, runner=None):
```
Comment (chhluo) [inline tools/analysis_tools/analyze_results.py:83]: update docstring

File: tools/analysis_tools/analyze_results.py [hunk 19]
```diff
@@ -73,28 +78,22 @@ class ResultVisualizer:
...
+        self.runner = runner
+        self.evaluator = runner.test_evaluator
 
     def _save_image_gts_results(self,
                                 dataset,
                                 results,
                                 performances,
-                                out_dir=None):
+                                out_dir=None,
+                                task='det'):
```
Comment (chhluo) [inline tools/analysis_tools/analyze_results.py:96]: update docstring

File: tools/analysis_tools/analyze_results.py [hunk 20]
```diff
@@ -157,25 +175,32 @@ def evaluate_and_show(self,
             eval_fn (callable, optional): Eval function, Default: None.
         """
 
+        self.visualizer.dataset_meta = dataset.metainfo
```
Comment (chhluo) [inline tools/analysis_tools/analyze_results.py:178]: update docstring

File: tools/analysis_tools/analyze_results.py [hunk 21]
```diff
@@ -105,6 +106,7 @@ def _save_image_gts_results(self,
                 in dataset and model's performance on them.
             out_dir (str, optional): The filename to write the image.
                 Defaults: None.
+            task (str): The task to be performed. Defaults: det
```
Comment (chhluo) [inline tools/analysis_tools/analyze_results.py:109]: ‘det’

**Review by chhluo — COMMENTED** [review]
LGTM

**Review by RangiLyu — APPROVED** [review]

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**